PerimeterXModule/Internals/Helpers/RemoteServer.cs

```diff
@@ -0,0 +1,194 @@
+
+			// Create a request to the server
+			HttpWebRequest request = (HttpWebRequest)WebRequest.Create(_remoteUrl);
+
+						// Set some options
+			var requestHeaders = _context.Request.Headers;
+			var proxyHeaders = new WebHeaderCollection();
+
+			foreach (string headerName in requestHeaders)
+			{
+				switch (headerName)
+				{
+					case "Connection":
+					case "Content-Length":
+					case "Date":
+					case "Expect":
+					case "Host":
+					case "If-Modified-Since":
+					case "Range":
+					case "Transfer-Encoding":
+					case "Proxy-Connection":
+						// Let IIS handle these
+					case "Accept":
+					case "Content-Type":
+					case "Referer":
+					case "User-Agent":
+						// Should be set after POST
+						break;
+
+					default:
+						proxyHeaders.Add(headerName, requestHeaders.Get(headerName));
+						break;
+				}
+			}
+
+			request.Headers = proxyHeaders;
+			request.KeepAlive = true;
+
+			request.Method = _context.Request.HttpMethod;
+			request.UserAgent = _context.Request.UserAgent;
+			request.Referer = _context.Request.Headers.Get("Referer");
+			request.Accept = _context.Request.ContentType;
+			request.Accept = _context.Request.Headers.Get("Accept");
+
+			// For POST, write the post data extracted from the incoming request
+			if (request.Method == "POST")
+			{
+				Stream clientStream = _context.Request.InputStream;
+				byte[] clientPostData = new byte[_context.Request.InputStream.Length];
+				clientStream.Read(clientPostData, 0, (int)_context.Request.InputStream.Length);
+
+				request.ContentType = _context.Request.ContentType;
+				request.ContentLength = clientPostData.Length;
+				Stream stream = request.GetRequestStream();
+				stream.Write(clientPostData, 0, clientPostData.Length);
+				stream.Close();
+			}
+
+			return request;
+
+		}
+
+		/// <summary>
+		/// Send the request to the remote server and return the response
+		/// </summary>
+		/// <param name="request">Request to send to the server </param>
+		/// <returns>Response received from the remote server
+		///           or null if page not found </returns>
+		public HttpWebResponse GetResponse(HttpWebRequest request)
+		{
+			HttpWebResponse response;
+
+			try
+			{
+				response = (HttpWebResponse)request.GetResponse();
+			}
+			catch (WebException e)
+			{
+				Debug.WriteLine("Failed to get response: " + e.Message, PxConstants.LOG_CATEGORY);
+				// Send 404 to client 
+				_context.Response.StatusCode = 404;
+				_context.Response.StatusDescription = "Page Not Found";
+				_context.Response.Write("Page not found");
+				return null;
+			}
+
+			return response;
+		}
+
+		/// <summary>
+		/// Return the response in bytes array format
+		/// </summary>
+		/// <param name="response">Response received
+		///             from the remote server </param>
+		/// <returns>Response in bytes </returns>
+		public byte[] GetResponseStreamBytes(HttpWebResponse response)
+		{
+			int bufferSize = 256;
```

Comment:
Is that some sort of a "default chunk size" value for reading from a stream? Let's have a const to hold it, and let's verify that it has to be 256 or can be bigger.

Reply:
👍